contracts/protocol/modules/v1/MorphoLeverageModule.sol

```diff
@@ -0,0 +1,920 @@
+     * @param _setTokenQuantity     Quantity of SetToken
+     * @param _component            Address of component
+     */
+    function componentRedeemHook(
+        ISetToken _setToken,
+        uint256 _setTokenQuantity,
+        IERC20 _component,
+        bool _isEquity
+    )
+        external
+        override
+        onlyModule(_setToken)
+    {
+        IMorpho.MarketParams memory setMarketParams = marketParams[_setToken];
+        if (_isEquity && setMarketParams.collateralToken == address(_component)) {
+            int256 componentCollateral = _setToken.getExternalPositionRealUnit(address(_component), address(this));
+            require(componentCollateral > 0, "Component must be negative");
+            uint256 notionalCollateral = componentCollateral.toUint256().preciseMul(_setTokenQuantity);
+            _withdraw(_setToken, setMarketParams, notionalCollateral);
+        }
+        if(!_isEquity) {
+            require(setMarketParams.loanToken == address(_component), "Debt component mismatch");
+            int256 componentDebt = _setToken.getExternalPositionRealUnit(address(_component), address(this));
+
+            require(componentDebt < 0, "Component must be negative");
+
+            uint256 notionalDebt = componentDebt.mul(-1).toUint256().preciseMul(_setTokenQuantity);
+            _repayBorrow(_setToken, setMarketParams, notionalDebt, 0);
+        }
+    }
+
+
+    /* ============ Internal Functions ============ */
+
+    /**
+     * @dev Reads outstanding borrow token debt and collateral token balance from Morpho
+     * and normalizes them  by the sets total supply to get the per token equivalent (position)
+     */
+    function _getCollateralAndBorrowPositions(
+        ISetToken _setToken,
+        IMorpho.MarketParams memory _marketParams,
+        uint256 _setTotalSupply
+    )
+        internal
+        view
+        returns(int256 collateralPosition, int256 borrowPosition)
+    {
+        (uint256 collateralBalance, uint256 borrowBalance, ) = _getCollateralAndBorrowBalances(_setToken, _marketParams);
+        collateralPosition = collateralBalance.preciseDiv(_setTotalSupply).toInt256();
+        borrowPosition = borrowBalance.preciseDivCeil(_setTotalSupply).toInt256().mul(-1);
+    }
+
+    /**
+     * @dev Reads outstanding borrow token debt and collateral token balance from Morpho
+     */
+    function _getCollateralAndBorrowBalances(
+        ISetToken _setToken,
+        IMorpho.MarketParams memory _marketParams
+    )
+        internal
+        view 
+        returns(uint256 collateralBalance, uint256 borrowBalance, uint256 borrowSharesU256)
+    {
+        // TODO: Review if it is more efficient to generate this id once and save in storage
+        bytes32 marketId = _marketParams.id();
+        ( , , uint128 totalBorrowAssets, uint128 totalBorrowShares, , ) = morpho.market(marketId);
+        ( , uint128 borrowShares, uint128 collateral) = morpho.position(marketId, address(_setToken));
```

Comment:
To get up to date balances you can use this library instead: https://github.com/morpho-org/morpho-blue/blob/3f018087e024538486858e87499a10f6283a9528/src/libraries/periphery/MorphoBalancesLib.sol#L109

Else the balance won't take into account the accrued interest, unless interest has been accrued already by triggering the `accrueInterest` function as done [here](https://github.com/IndexCoop/index-protocol/pull/41/files#r1712614077).